shared/reducers/fs.js

```diff
@@ -31,31 +32,40 @@ export default function(state: Types.State = initialState, action: FsGen.Actions
 
         if (item.type !== 'folder') return item
         if (!original || original.type !== 'folder') return item
-        if (original.progress === 'loaded' && item.progress === 'pending') {
-          // Don't override a loaded item into pending. This is specifically
-          // for the case where user goes back out of a folder where we could
-          // override the folder into an empty one. With this, next user
-          // navigates into the folder they would see the old list (instead of
-          // placeholder), which then gets updated when we hear back from RPC.
-          return original
+
+        // Make flow happy by referencing them with a new name that's
+        // explicitly typed.
+        const originalFolder: Types.FolderPathItem = original
+        let newItem: Types.FolderPathItem = item
+
+        if (originalFolder.progress === 'loaded' && newItem.progress === 'pending') {
+          // We don't want to override a loaded folder into pending, because
+          // otherwise next time user goes into that folder we'd show
+          // placeholders. We also don't want to simply use the original
+          // PathItem, since it's possible some metadata has updated. So use
+          // the new item, but reuse children, progress, and tlfMeta fields.
+          if (originalFolder.type === 'folder' && newItem.type === 'folder') {
+            // make flow happy
+            newItem = newItem.set('children', originalFolder.children).set('progress', 'loaded')
```

Comment:
What about original's other properties?

Reply:
Good catch. I think this is missing tlfMeta, etc.